### ISSUE TITLE
3x write speed increase when writing a dummy buffer before flush

### ISSUE THREAD

**Issue by Crunkle** [netty/netty#11110]
### Description
I have discovered that writing a dummy buffer immediately before `ctx.flush()` will heavily improve write performance on Windows systems. I have tested this and benchmarked on two separate devices receiving similar results - dummy write is much faster.

```
A: 125506200ns (dummy write before flush)
B: 351542000ns (no dummy write)
```

I have not tested whether this is reproducible on epoll and kqueue backends so this may be isolated to Windows.

### Steps to reproduce
Minimal working example, tested using NIO under Windows: https://pastebin.com/4Eb975Qw

### Minimal yet complete reproducer code (or URL to code)
```java
Channel channel = new NioDatagramChannel()
    .pipeline().addLast(new ChannelOutboundHandlerAdapter() {

        @Override
        public void flush(ChannelHandlerContext ctx) {
            if (flushWrite) {
                // Writing a dummy buffer before flush improves performance
                ctx.write(Unpooled.copyInt(0));
            }

            ctx.flush();
        }
    })
    .channel();

eventLoopGroup
    .register(channel).sync().channel()
    .connect(new InetSocketAddress("127.0.0.1", 1234)).sync();

long t = System.nanoTime();
for (int i = 0; i < 5000; i++) {
    channel.writeAndFlush(Unpooled.wrappedBuffer(new byte[1024])).sync();
}
System.out.println((flushWrite ? "A: " : "B: ") + (System.nanoTime() - t) + "ns");

channel.close();
```

### Netty version
`4.1.60.Final`

### JVM version (e.g. `java -version`)
`Oracle JRE 1.8.0_181-b13`

### OS version (e.g. `uname -a`)
`Windows 10 19042`

**Comment by chrisvest** [comment]
Note that `Unpooled.copyInt(0)` creates a buffer with 4 readable zero-bytes in it; read-index 0, write-index 4. So it's not exactly a dummy buffer. The `Unpooled.EMPTY_BUFFER` is, though.

**Comment by Crunkle** [comment]
> Note that `Unpooled.copyInt(0)` creates a buffer with 4 readable zero-bytes in it; read-index 0, write-index 4. So it's not exactly a dummy buffer. The `Unpooled.EMPTY_BUFFER` is, though.

Indeed, I simply call it a dummy buffer as it is unused; this is not a viable solution for production code. Writing an empty buffer does not reproduce the speed increase, it seems to require any non-zero buffer.

**Comment by normanmaurer** [comment]
I wonder if this is just the effect of TCP_NODELAY and writing enough to finally send the buffer. Can you experiment with setting `TCP_NODELAY` to true ?

**Comment by Crunkle** [comment]
Just tried this and the issue persists. I believe `TCP_NODELAY` is ignored on UDP sockets but worth a shot. I switched over to TCP and the issue does not occur regardless of any channel options set (so can conclude this is isolated to NIO UDP only).

**Comment by chrisvest** [comment]
Try running on a newer JVM (11 or better), and profile it with Java Mission Control.

On MacOS I also see a difference, where writing the "dummy" buffer is faster by a few percent, though not a 3x difference.
When the extra write is there, the NioEventLoop goes into `processSelectedKeys`, while without it there's no such call and somehow that's slower. I'm not sure why this makes such a difference.

Note that using IO buffers instead of heap buffers will speed up the program in both cases, because it removes a need to copy memory:

```patch
- channel.writeAndFlush(Unpooled.wrappedBuffer(new byte[1024])).sync();
+ channel.writeAndFlush(channel.alloc().ioBuffer(1024)).sync();
```

**Comment by Crunkle** [comment]
Thanks for testing. I have a minor improvement using direct buffers but nothing substantial under Windows. Just want to note that in your diff the first buffer is 1024 in length and the second is 1024 in capacity, not length.
